File: flowquery-py/src/parsing/functions/to_integer.py
"""ToInteger function."""

from typing import Any

from .function import Function
from .function_metadata import FunctionDef


@FunctionDef({
    "description": "Converts a value to an integer",
    "category": "scalar",
    "parameters": [
        {"name": "value", "description": "The value to convert to an integer", "type": "any"}
    ],
    "output": {"description": "The integer representation of the value", "type": "number", "example": 42},
    "examples": [
        'RETURN toInteger("42")',
        "RETURN toInteger(3.14)",
        "RETURN toInteger(true)"
    ]
})
class ToInteger(Function):
    """ToInteger function.

    Converts a value to an integer.
    """

    def __init__(self) -> None:
        super().__init__("tointeger")
        self._expected_parameter_count = 1

    def value(self) -> Any:
        val = self.get_children()[0].value()
        if val is None:
            return None
        if isinstance(val, bool):
            return 1 if val else 0
        if isinstance(val, (int, float)):
            return int(val)
        if isinstance(val, str):
            trimmed = val.strip()
            try:
                return int(float(trimmed))
            except (ValueError, OverflowError):
                raise ValueError(f'Cannot convert string "{val}" to integer')
        raise ValueError("toInteger() expects a number, string, or boolean")

File: src/parsing/functions/to_integer.ts
import Function from "./function";
import { FunctionDef } from "./function_metadata";

@FunctionDef({
    description: "Converts a value to an integer",
    category: "scalar",
    parameters: [
        {
            name: "value",
            description: "The value to convert to an integer",
            type: "any",
        },
    ],
    output: {
        description: "The integer representation of the value",
        type: "number",
        example: "42",
    },
    examples: ['RETURN toInteger("42")', "RETURN toInteger(3.14)", "RETURN toInteger(true)"],
})
class ToInteger extends Function {
    constructor() {
        super("tointeger");
        this._expectedParameterCount = 1;
    }

    public value(): any {
        const val = this.getChildren()[0].value();
        if (val === null || val === undefined) {
            return null;
        }
        if (typeof val === "boolean") {
            return val ? 1 : 0;
        }
        if (typeof val === "number") {
            return Math.trunc(val);
        }
        if (typeof val === "string") {
            const trimmed = val.trim();
            const parsed = Number(trimmed);
            if (isNaN(parsed)) {
                throw new Error(`Cannot convert string "${val}" to integer`);
            }
            return Math.trunc(parsed);
        }
        throw new Error("toInteger() expects a number, string, or boolean");
    }
}

export default ToInteger;
